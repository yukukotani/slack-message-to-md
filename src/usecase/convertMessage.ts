import { parseAttachments } from "../libs/attachmentParser";
import { parseBlocks } from "../libs/blockParser";
import {
  createUnknownError,
  handleMissingContent,
  wrapSafeExecution,
} from "../libs/errorHandler";
import { parseFiles } from "../libs/fileParser";
import {
  formatEditedInfo,
  formatReactions,
  formatUserHeader,
} from "../libs/metadataParser";
import { formatMrkdwn } from "../libs/textFormatter";
import type {
  ConversionResult,
  SlackMessage,
  UserMapping,
} from "../libs/types";
import { hasAttachments, hasBlocks, hasFiles } from "../libs/types";
import { SlackMessageSchema } from "../libs/validation";

export function convertMessage(
  message: SlackMessage,
  userMapping?: UserMapping,
): ConversionResult {
  try {
    // コンテンツがあるかチェック
    if (!hasContent(message)) {
      return handleMissingContent(message);
    }

    const sections: string[] = [];

    // 1. ヘッダー（ユーザー + タイムスタンプ）
    const header = wrapSafeExecution(
      () => formatUserHeader(message.user, message.ts, userMapping),
      "",
    );
    if (header) {
      sections.push(header);
    }

    // 2. メインコンテンツ（blocks が優先、なければ text）
    if (hasBlocks(message)) {
      const blocksContent = wrapSafeExecution(
        () => parseBlocks(message.blocks, userMapping),
        "",
      );
      if (blocksContent) {
        sections.push(blocksContent);
      }
    } else if (message.text) {
      const textContent = wrapSafeExecution(
        () => formatMrkdwn(message.text || "", userMapping),
        message.text || "",
      );
      if (textContent) {
        sections.push(textContent);
      }
    }

    // 3. アタッチメント
    if (hasAttachments(message)) {
      const attachmentContent = wrapSafeExecution(
        () => parseAttachments(message.attachments, userMapping),
        "",
      );
      if (attachmentContent) {
        sections.push(attachmentContent);
      }
    }

    // 4. ファイル
    if (hasFiles(message)) {
      const fileContent = wrapSafeExecution(
        () => parseFiles(message.files),
        "",
      );
      if (fileContent) {
        sections.push(fileContent);
      }
    }

    // 5. リアクション
    if (message.reactions && message.reactions.length > 0) {
      const reactionsContent = wrapSafeExecution(
        () => formatReactions(message.reactions),
        "",
      );
      if (reactionsContent) {
        sections.push(reactionsContent);
      }
    }

    // 6. 編集情報
    if (message.edited) {
      const editedInfo = wrapSafeExecution(
        () => formatEditedInfo(message.edited, userMapping),
        "",
      );
      if (editedInfo) {
        sections.push(editedInfo);
      }
    }

    // 結合
    const markdown = sections
      .filter((section) => section.length > 0)
      .join("\n\n");

    if (markdown.length === 0) {
      return handleMissingContent(message);
    }

    return {
      success: true,
      markdown,
    };
  } catch (error) {
    return createUnknownError(error);
  }
}

export function convertMultipleMessages(
  messages: SlackMessage[],
  userMapping?: UserMapping,
): ConversionResult[] {
  return messages.map((message) => convertMessage(message, userMapping));
}

export function convertMessageWithValidation(
  message: unknown,
  userMapping?: UserMapping,
): ConversionResult {
  const parseResult = SlackMessageSchema.safeParse(message);

  if (!parseResult.success) {
    return {
      success: false,
      error: {
        code: "INVALID_JSON",
        message: "メッセージの形式が正しくありません",
        details: parseResult.error.format(),
      },
    };
  }

  return convertMessage(parseResult.data as SlackMessage, userMapping);
}

function hasContent(message: SlackMessage): boolean {
  return !!(
    message.text ||
    hasBlocks(message) ||
    hasAttachments(message) ||
    hasFiles(message) ||
    (message.reactions && message.reactions.length > 0) ||
    message.reply_count ||
    message.edited
  );
}
