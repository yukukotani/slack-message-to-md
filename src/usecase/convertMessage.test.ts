import { describe, expect, it } from "vitest";
import type { SlackMessage } from "../libs/types";
import { convertMessage, convertMultipleMessages } from "./convertMessage";

describe("convertMessage", () => {
  it("基本的なテキストメッセージを変換", () => {
    const message: SlackMessage = {
      type: "message",
      text: "Hello *world*!",
      user: "U123456",
      ts: "1704980400",
    };
    const result = convertMessage(message);
    expect(result).toMatchSnapshot();
  });

  it("blocksを含むメッセージを変換", () => {
    const message: SlackMessage = {
      type: "message",
      user: "U123456",
      ts: "1704980400",
      blocks: [
        {
          type: "header",
          text: { type: "plain_text", text: "Important Notice" },
        },
        {
          type: "section",
          text: {
            type: "mrkdwn",
            text: "This is a *section* with formatting.",
          },
        },
      ],
    };
    const result = convertMessage(message);
    expect(result).toMatchSnapshot();
  });

  it("attachmentsを含むメッセージを変換", () => {
    const message: SlackMessage = {
      type: "message",
      text: "Check this out:",
      user: "U123456",
      ts: "1704980400",
      attachments: [
        {
          title: "Sample Attachment",
          text: "Attachment content",
          color: "good",
        },
      ],
    };
    const result = convertMessage(message);
    expect(result).toMatchSnapshot();
  });

  it("filesを含むメッセージを変換", () => {
    const message: SlackMessage = {
      type: "message",
      text: "Uploaded a file:",
      user: "U123456",
      ts: "1704980400",
      files: [
        {
          id: "F123456",
          name: "document.pdf",
          mimetype: "application/pdf",
          url_private:
            "https://files.slack.com/files-pri/T123456-F123456/document.pdf",
          size: 1048576,
        },
      ],
    };
    const result = convertMessage(message);
    expect(result).toMatchSnapshot();
  });

  it("reactionsを含むメッセージを変換", () => {
    const message: SlackMessage = {
      type: "message",
      text: "Great job!",
      user: "U123456",
      ts: "1704980400",
      reactions: [
        {
          name: "thumbsup",
          users: ["U234567", "U345678"],
          count: 2,
        },
        {
          name: "heart",
          users: ["U456789"],
          count: 1,
        },
      ],
    };
    const result = convertMessage(message);
    expect(result).toMatchSnapshot();
  });

  it("スレッド情報を含むメッセージを変換", () => {
    const message: SlackMessage = {
      type: "message",
      text: "Original message",
      user: "U123456",
      ts: "1704980400",
      reply_count: 3,
      reply_users: ["U234567", "U345678"],
      reply_users_count: 2,
    };
    const result = convertMessage(message);
    expect(result).toMatchSnapshot();
  });

  it("編集済みメッセージを変換", () => {
    const message: SlackMessage = {
      type: "message",
      text: "Edited message",
      user: "U123456",
      ts: "1704980400",
      edited: {
        user: "U123456",
        ts: "1704984000",
      },
    };
    const result = convertMessage(message);
    expect(result).toMatchSnapshot();
  });

  it("複数要素を含む複雑なメッセージを変換", () => {
    const message: SlackMessage = {
      type: "message",
      text: "Main message text",
      user: "U123456",
      ts: "1704980400",
      blocks: [
        {
          type: "section",
          text: { type: "plain_text", text: "Block content" },
        },
      ],
      attachments: [
        {
          title: "Attachment",
          text: "Attachment text",
        },
      ],
      files: [
        {
          id: "F123456",
          name: "image.png",
          mimetype: "image/png",
          url_private:
            "https://files.slack.com/files-pri/T123456-F123456/image.png",
          size: 2048,
        },
      ],
      reactions: [
        {
          name: "thumbsup",
          users: ["U234567"],
          count: 1,
        },
      ],
    };
    const result = convertMessage(message);
    expect(result).toMatchSnapshot();
  });

  it("コンテンツがない場合のエラーハンドリング", () => {
    const message: SlackMessage = {
      type: "message",
      ts: "1704980400",
    };
    const result = convertMessage(message);
    expect(result).toMatchSnapshot();
  });
});

describe("convertMultipleMessages", () => {
  it("複数メッセージを順番に変換", () => {
    const messages: SlackMessage[] = [
      {
        type: "message",
        text: "First message",
        user: "U123456",
        ts: "1704980400",
      },
      {
        type: "message",
        text: "Second message",
        user: "U234567",
        ts: "1704984000",
      },
    ];
    const results = convertMultipleMessages(messages);
    expect(results).toMatchSnapshot();
  });

  it("エラーメッセージも含めて処理", () => {
    const messages: SlackMessage[] = [
      {
        type: "message",
        text: "Valid message",
        user: "U123456",
        ts: "1704980400",
      },
      {
        type: "message",
        ts: "1704984000", // テキストなし
      },
    ];
    const results = convertMultipleMessages(messages);
    expect(results).toMatchSnapshot();
  });

  it("空の配列を処理", () => {
    const results = convertMultipleMessages([]);
    expect(results).toMatchSnapshot();
  });
});
