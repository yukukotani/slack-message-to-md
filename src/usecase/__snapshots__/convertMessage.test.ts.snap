// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`convertMessage 基本的なテキストメッセージを変換 1`] = `
{
  "markdown": 
"**@U123456** - 2024-01-11 13:40:00

Hello **world**!"
,
  "success": true,
}
`;

exports[`convertMessage blocksを含むメッセージを変換 1`] = `
{
  "markdown": 
"**@U123456** - 2024-01-11 13:40:00

# Important Notice

<section>

This is a **section** with formatting.

</section>"
,
  "success": true,
}
`;

exports[`convertMessage attachmentsを含むメッセージを変換 1`] = `
{
  "markdown": 
"**@U123456** - 2024-01-11 13:40:00

Check this out:

<attachment>
**Sample Attachment**

Attachment content
</attachment>"
,
  "success": true,
}
`;

exports[`convertMessage filesを含むメッセージを変換 1`] = `
{
  "markdown": 
"**@U123456** - 2024-01-11 13:40:00

<file id="F123456" mimetype="application/pdf" href="https://files.slack.com/files-pri/T123456-F123456/document.pdf" size="1048576">document.pdf</file>"
,
  "success": true,
}
`;

exports[`convertMessage reactionsを含むメッセージを変換 1`] = `
{
  "markdown": 
"**@U123456** - 2024-01-11 13:40:00

Great job!

<reaction count="2" users="U234567 U345678">thumbsup</reaction>
<reaction count="1" users="U456789">heart</reaction>"
,
  "success": true,
}
`;

exports[`convertMessage 複数要素を含む複雑なメッセージを変換 1`] = `
{
  "markdown": 
"**@U123456** - 2024-01-11 13:40:00

<section>

Block content

</section>

<attachment>
**Attachment**

Attachment text
</attachment>

<file id="F123456" mimetype="image/png" href="https://files.slack.com/files-pri/T123456-F123456/image.png" size="2048">image.png</file>

<reaction count="1" users="U234567">thumbsup</reaction>"
,
  "success": true,
}
`;

exports[`convertMessage コンテンツがない場合のエラーハンドリング 1`] = `
{
  "error": {
    "code": "MISSING_CONTENT",
    "details": {
      "ts": "1704980400",
      "type": "message",
    },
    "message": "No content to convert",
  },
  "success": false,
}
`;

exports[`convertMultipleMessages 複数メッセージを順番に変換 1`] = `
[
  {
    "markdown": 
"**@U123456** - 2024-01-11 13:40:00

First message"
,
    "success": true,
  },
  {
    "markdown": 
"**@U234567** - 2024-01-11 14:40:00

Second message"
,
    "success": true,
  },
]
`;

exports[`convertMultipleMessages エラーメッセージも含めて処理 1`] = `
[
  {
    "markdown": 
"**@U123456** - 2024-01-11 13:40:00

Valid message"
,
    "success": true,
  },
  {
    "error": {
      "code": "MISSING_CONTENT",
      "details": {
        "ts": "1704984000",
        "type": "message",
      },
      "message": "No content to convert",
    },
    "success": false,
  },
]
`;

exports[`convertMultipleMessages 空の配列を処理 1`] = `[]`;
